File: eco-api/src/main/java/com/willfp/eco/util/ListUtils.java
package com.willfp.eco.util;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Utilities / API methods for lists.
 */
public final class ListUtils {
    /**
     * Initialize 2D list of a given size.
     *
     * @param rows    The amount of rows.
     * @param columns The amount of columns.
     * @param <T>     The type of the object stored in the list.
     * @return The list, filled will null objects.
     */
    @NotNull
    public static <T> List<List<T>> create2DList(final int rows,
                                                 final int columns) {
        List<List<T>> list = new ArrayList<>(rows);
        while (list.size() < rows) {
            List<T> row = new ArrayList<>(columns);
            while (row.size() < columns) {
                row.add(null);
            }
            list.add(row);
        }

        return list;
    }

    /**
     * Convert a list potentially containing duplicates to a map where the value is the frequency of the key.
     *
     * @param list The list.
     * @param <T>  The type parameter of the list.
     * @return The frequency map.
     */
    @NotNull
    public static <T> Map<T, Integer> listToFrequencyMap(@NotNull final List<T> list) {
        Map<T, Integer> frequencyMap = new HashMap<>();
        for (T object : list) {
            if (frequencyMap.containsKey(object)) {
                frequencyMap.put(object, frequencyMap.get(object) + 1);
            } else {
                frequencyMap.put(object, 1);
            }
        }

        return frequencyMap;
    }

    /**
     * Convert nullable object to either singleton list or empty list.
     *
     * @param object The object.
     * @param <T>    The type of the object.
     * @return The list.
     */
    @NotNull
    public static <T> List<T> toSingletonList(@Nullable final T object) {
        if (object == null) {
            return Collections.emptyList();
        } else {
            return Collections.singletonList(object);
        }
    }

    /**
     * Get element from list or return null if out of bounds.
     *
     * @param list  The list.
     * @param index The index.
     * @param <T>   The type of the list.
     * @return The found element, or null if out of bounds.
     */
    @Nullable
    public static <T> T getOrNull(@Nullable final List<T> list,
                                  final int index) {
        if (list == null) {
            return null;
        }

        return index >= 0 && index < list.size() ? list.get(index) : null;
    }

    /**
     * Get if an iterable of strings contains a certain element regardless of case.
     *
     * @param list    The list.
     * @param element The element.
     * @return If contained.
     */
    public static boolean containsIgnoreCase(@NotNull final Iterable<String> list,
                                             @NotNull final String element) {
        for (String s : list) {
            if (s.equalsIgnoreCase(element)) {
                return true;
            }
        }

        return false;
    }

    private ListUtils() {
        throw new UnsupportedOperationException("This is a utility class and cannot be instantiated");
    }
}

File: eco-api/src/main/kotlin/com/willfp/eco/util/ListUtils.kt
@file:JvmName("ListUtilsExtensions")

package com.willfp.eco.util

/**
 * Convert this list, which may contain duplicates, into a map of each distinct element to
 * the number of times it occurs in the list.
 *
 * @return The frequency map.
 * @see ListUtils.listToFrequencyMap
 */
fun <T> List<T>.toFrequencyMap(): Map<T, Int> =
    ListUtils.listToFrequencyMap(this)

/**
 * Get if this iterable contains an element, ignoring case.
 *
 * @param element The element to look for.
 * @return If an element equal to it, ignoring case, is contained.
 * @see ListUtils.containsIgnoreCase
 */
fun Iterable<String>.containsIgnoreCase(element: String): Boolean =
    ListUtils.containsIgnoreCase(this, element)

/**
 * Create a mutable 2D list of a fixed size, where every cell is initialised to null.
 *
 * The outer list holds `rows` inner lists, each holding `columns` elements. As the cells
 * start out null, `T` should be a nullable type.
 *
 * @param rows    The number of rows (the size of the outer list).
 * @param columns The number of columns (the size of each inner list).
 * @return The created list, filled with nulls.
 * @see ListUtils.create2DList
 */
fun <T> create2DList(rows: Int, columns: Int): MutableList<MutableList<T>> =
    ListUtils.create2DList(rows, columns)

/**
 * Convert this nullable object into a list containing it.
 *
 * @return An immutable singleton list containing the receiver, or an immutable empty list if it is null.
 * @see ListUtils.toSingletonList
 */
fun <T> T?.toSingletonList(): List<T> =
    ListUtils.toSingletonList(this)
